sdk/offchain/src/identifier.rs

```diff
@@ -0,0 +1,386 @@
+            };
+        Ok(encode_intent(
+            &encoded_account,
+            self.currency_code.as_deref(),
+            self.amount,
+        ))
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum IntentIdentifierError {
+    #[error(transparent)]
+    AccountAddress(#[from] AccountAddressParseError),
+    #[error(transparent)]
+    Bech32(#[from] bech32::Error),
+    #[error("{0}")]
+    Parse(String),
+    #[error(transparent)]
+    Subaddress(#[from] SubaddressParseError),
+    #[error(transparent)]
+    Url(#[from] url::ParseError),
+}
+
+trait ToHRP {
+    /// Returns the hrp (human readable prefix) as defined in
+    /// https://dip.diem.com/dip-5/#format
+    fn to_hrp(&self) -> Result<String>;
+}
+
+impl ToHRP for ChainId {
+    fn to_hrp(&self) -> Result<String> {
+        let named_chain = NamedChain::from_chain_id(self);
+        let hrp = match named_chain {
+            Ok(NamedChain::MAINNET) => String::from("dm"),
+            Ok(NamedChain::PREMAINNET) => String::from("pdm"),
+            Ok(NamedChain::TESTNET) => String::from("tdm"),
+            Ok(NamedChain::DEVNET) => String::from("tdm"),
+            Ok(NamedChain::TESTING) => String::from("tdm"),
+            Err(error_str) => return Err(anyhow!(error_str)),
+        };
+        Ok(hrp)
+    }
+}
+
+/// Encode onchain address and subaddress with human readable prefix (hrp) into bech32 format.
+fn encode_account(
+    hrp: &str,
+    account_address: AccountAddress,
+    subaddress: Subaddress,
+) -> Result<String, bech32::Error> {
+    let five_bit_data = [account_address.to_vec(), subaddress.to_vec()]
+        .concat()
+        .to_base32();
+    let diem_encoding_version = vec![u5::try_from_u8(DIEM_BECH32_VERSION)?];
+    let bytes = [diem_encoding_version, five_bit_data].concat();
+    bech32::encode(hrp, bytes, bech32::Variant::Bech32)
+}
+
+/// Decodes an encoded address using bech32, ensuring a matching hrp (human readable prefix).
+fn decode_account(
+    encoded_address_option: Option<&str>,
+    expected_hrp: &str,
+) -> Result<(AccountAddress, Subaddress), IntentIdentifierError> {
+    if encoded_address_option.is_none() {
+        return Err(IntentIdentifierError::Parse(
+            "Encoded address is empty".to_string(),
+        ));
+    }
+    let encoded_address = encoded_address_option.unwrap();
+
+    let (hrp, data, _variant) = bech32::decode(encoded_address)?;
+
+    let diem_address_version = data[0].to_u8();
+    if diem_address_version != DIEM_BECH32_VERSION {
+        return Err(IntentIdentifierError::Parse(format!(
+            "Unsupported Diem Bech32 Version {}, expected {}",
+            diem_address_version, DIEM_BECH32_VERSION,
+        )));
+    }
+
+    let payload = Vec::<u8>::from_base32(&data[1..])?;
+    if expected_hrp != hrp {
+        return Err(IntentIdentifierError::Parse(format!(
+            "Unexpected HRP {}, expected {}",
+            hrp, expected_hrp,
+        )));
+    }
+
```

Comment:
probably should check payload length == account address len + subaddress len before parsing them.